pymobiledevice3/remote/tunnel_service.py

```diff
@@ -965,20 +968,38 @@ async def start_tunnel(
         raise Exception(f'Bad value for protocol_handler: {protocol_handler}')
 
 
-async def get_core_device_tunnel_services(bonjour_timeout: float = DEFAULT_BONJOUR_TIMEOUT) \
-        -> List[CoreDeviceTunnelService]:
+async def get_core_device_tunnel_services(bonjour_timeout: float = DEFAULT_BONJOUR_TIMEOUT,
+        udid: Optional[str] = None) -> List[CoreDeviceTunnelService]:
     result = []
-    for rsd in await get_rsds(bonjour_timeout=bonjour_timeout):
-        result.append(create_core_device_tunnel_service_using_rsd(rsd))
+    invalid_service_error = None
+    for rsd in await get_rsds(bonjour_timeout=bonjour_timeout, udid=udid):
+        if udid is None and Version(rsd.product_version) < Version('17.0'):
```

Comment:
technically it's possible to access ios16 devices too, eventhough the exception is raised, so i rather we'll just catch the exception instead

Reply:
Note 1: you can *force* pymobiledevice3 to attempt connections to 16.x devices with "--udid".  So it remains possibly to try(the version check is short-circuted by an is-not-none udid).  They just won't "autodetect" and appear on the multiple-choice list.

The problem is that some of the iOS16.x exceptions are so bad that they make the start-tunnel command very hard to use for "good" devices.  For the "pairing" exception mentioned in my ticket, you must acknowledge the pairing dialog on the bad device before pymobiledevice3 will continue.  Until then, pymobiledevice3 hangs.  But, then the pairing fails anyway, so the next time you run start-tunnel, that device will be triggered to pair again.  Even if we caught the pairing exception, you would have to physically tap the screen on every 'bad' iOS16 device each time you run start-tunnel. 

The second problem is network timeouts.  Even if we catch a network timeout exception, the timeout has to happen first.  If I have 3 "bad" iOS16 devices attached with timeouts, start-tunnel will hang for 3+ minutes before producing any output, waiting to catch the 3 network timeout exceptions.

In truth, my use of "--udid" works around this for me.  But you'll have end-users who don't understand why "start-tunnel" hangs, just because they happen to have iOS16 devices attached.  Very hard to diagnose.  They'll report bugs "pymobiledevice3 hangs...", and the only workaround will be to unplug the offending devices.

iOS16.x is currently so broken with RSD that I think you would face fewer complaints and less confusion from users if you just ignored them by default.  They rarely seem to work with RSD anyway.  If iOS 16.7 or 16.8 are known to work, I'd be more in favor of a " < Version('16.7')" check, than throwing the version check away.

If you insist, I'll just catch generic exceptions as you've suggested.  But like I said, that still leaves significant problems when using pymobiledevice3 with iOS17 and iOS16 devices on the same host.

Reply:
I see. this bug will remain till I refactor it'll all to be pure-asyncio.
Very well then